### PR TITLE
[socketcluster-server][ag-auth] Fix dependency & add its type definition

### PR DESCRIPTION
**Pull request by DanielRose**
- [x] Use a meaningful title for the pull request. Include the name of the package modified.
- [x] Test the change in your own code. (Compile and run.)
- [x] [Add or edit tests](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#my-package-teststs) to reflect the change.
- [x] Follow the advice from the [readme](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#make-a-pull-request).
- [x] Avoid [common mistakes](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#common-mistakes).
- [x] [Run `npm test <package to test>`](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#running-tests).

For ag-auth:

- [x] The package does not already provide its own types, or cannot have its `.d.ts` files generated via `--declaration`
- [x] If this is for an npm package, match the name. If not, do not conflict with the name of an npm package.
- [x] Create it with `dts-gen --dt`, not by basing it on an existing project.
- [x] Represents shape of module/library [correctly](https://www.typescriptlang.org/docs/handbook/declaration-files/library-structures.html)
- [x] `tslint.json` [should contain](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#linter-tslintjson) `{ "extends": "dtslint/dt.json" }`, and no additional rules.
- [x] `tsconfig.json` [should have](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#tsconfigjson) `noImplicitAny`, `noImplicitThis`, `strictNullChecks`, and `strictFunctionTypes` set to `true`.

For socketcluster-server

- [x] Provide a URL to documentation or source code which provides context for the suggested changes: https://github.com/SocketCluster/socketcluster-server/blob/master/server.js#L2

### CODE REVIEW

File: types/ag-auth/index.d.ts [hunk 1]
```diff
@@ -0,0 +1,23 @@
...
+// Project: https://github.com/SocketCluster/ag-auth
+// Definitions by: Daniel Rose <https://github.com/DanielRose>
+// Definitions: https://github.com/DefinitelyTyped/DefinitelyTyped
+// Minimum TypeScript Version: 3.7
+
+import { VerifyOptions, Secret, SignOptions, Jwt } from 'jsonwebtoken';
+
+declare class AuthEngine implements AuthEngine.AGAuthEngine {
+    constructor();
+
```
Comment (peterblazejewicz) [inline types/ag-auth/index.d.ts:11]:
```suggestion
```
parameterless are always there

File: types/ag-auth/index.d.ts [hunk 2]
```diff
@@ -0,0 +1,23 @@
...
+    verifyToken(signedToken: string | null, key: Secret, options?: VerifyOptions): Promise<Jwt>;
+    signToken(token: string | object | Buffer, key: Secret, options?: SignOptions): Promise<string | undefined>;
+}
+
+export = AuthEngine;
+
+declare namespace AuthEngine {
```
Comment (peterblazejewicz) [inline types/ag-auth/index.d.ts:18]: why do you need those? The package just declares a class like constructor and exports it IMO:
https://github.com/SocketCluster/ag-auth/blob/master/index.js#L6

File: types/socketcluster-server/socketcluster-server-tests.ts [hunk 3]
```diff
@@ -1,7 +1,10 @@
 // Adapted from README
 
 import http = require('http');
+import agAuth = require('ag-auth');
```
Comment (peterblazejewicz) [inline types/socketcluster-server/socketcluster-server-tests.ts:4]: this should most probably read:
```ts
import AuthEngine = require('ag-auth');
``` 
https://github.com/SocketCluster/socketcluster-server/blob/ea2175a4d8548ca14bd9b5e7ad2cd3e94d70bc40/server.js#L2

File: types/socketcluster-server/socketcluster-server-tests.ts [hunk 4]
```diff
@@ -81,3 +84,31 @@ agServer.setMiddleware(agServer.MIDDLEWARE_INBOUND, async middlewareStream => {
...
+agServer = socketClusterServer.attach(httpServer, {
+    wsEngine: require('ws')
+});
+
+agServer = socketClusterServer.attach(httpServer, {
+    wsEngine: 'ws'
+});
+
+agServer = socketClusterServer.attach(httpServer, {
+    authEngine: new agAuth()
```
Comment (peterblazejewicz) [inline types/socketcluster-server/socketcluster-server-tests.ts:99]: so, if we use proper naming:
```suggestion
    authEngine: new AuthEngine()
```
https://github.com/SocketCluster/socketcluster-server/blob/ea2175a4d8548ca14bd9b5e7ad2cd3e94d70bc40/server.js#L135

File: types/ag-auth/index.d.ts (hunk 2 above)
Comment (peterblazejewicz) [reply types/ag-auth/index.d.ts:18]: One can have this 'AuthEngine' contract within declarin options module (socker server). In TS the class here can be just same shape, but wihtout hard contract between modules

File: types/ag-auth/index.d.ts [hunk 5]
```diff
@@ -0,0 +1,23 @@
...
+    interface AGAuthEngine {
+        verifyToken(signedToken: string | null, key: Secret, options?: VerifyOptions): Promise<Jwt>;
+        signToken(token: string | object | Buffer, key: Secret, options?: SignOptions): Promise<string | undefined>;
+    }
```
Comment (peterblazejewicz) [inline types/ag-auth/index.d.ts:22]: you can remove this completely, check in other place what you can do

**Comment by peterblazejewicz** [comment]
you can derive a type, required in server to accept object conforming the interface in `options` and in `setAuth` by doing this:
```ts
type AuthEngineType = Pick<AuthEngine, "verifyToken" | "signToken">;
```
that way you will avoid duplication and tight, direct coupling between those modules (requirement to define an interface in ag-auth)

**Review by peterblazejewicz — APPROVED** [review]
LGTM!

**Comment by DanielRose** [comment]
Ready to merge

**Pull request merged**